spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala

```diff
@@ -0,0 +1,224 @@
+    }
+
+    // assign default values to optional params
+    params.foreach {
+      case p if !p.required && !nameToArgMap.contains(p.name) =>
+        val position = nameToPositionMap(p.name)
+        values(position) = p.defaultValue
+      case _ =>
+    }
+
+    values
+  }
+
+  private def buildNameToArgMap(
+      params: Seq[ProcedureParameter],
+      args: Seq[CallArgument],
+      nameToPositionMap: Map[String, Int]): Map[String, CallArgument] = {
+
+    val containsNamedArg = args.exists(_.isInstanceOf[NamedArgument])
+    val containsPositionalArg = args.exists(_.isInstanceOf[PositionalArgument])
+
+    if (containsNamedArg && containsPositionalArg) {
+      throw new AnalysisException("Named and positional arguments cannot be mixed")
+    }
+
+    if (containsNamedArg) {
+      buildNameToArgMapUsingNames(args, nameToPositionMap)
+    } else {
+      buildNameToArgMapUsingPositions(args, params)
+    }
+  }
+
+  private def buildNameToArgMapUsingNames(
+      args: Seq[CallArgument],
+      nameToPositionMap: Map[String, Int]): Map[String, CallArgument] = {
+
+    val namedArgs = args.asInstanceOf[Seq[NamedArgument]]
+
+    val validationErrors = namedArgs.groupBy(_.name).collect {
+      case (name, matchingArgs) if matchingArgs.size > 1 => s"Duplicate procedure argument: $name"
+      case (name, _) if !nameToPositionMap.contains(name) => s"Unknown argument: $name"
+    }
+
+    if (validationErrors.nonEmpty) {
+      throw new AnalysisException(s"Could not build name to arg map: ${validationErrors.mkString(", ")}")
+    }
+
+    namedArgs.map(arg => arg.name -> arg).toMap
+  }
+
+  private def buildNameToArgMapUsingPositions(
+      args: Seq[CallArgument],
+      params: Seq[ProcedureParameter]): Map[String, CallArgument] = {
+
+    if (args.size > params.size) {
+      throw new AnalysisException("Too many arguments for procedure")
+    }
+
+    args.zipWithIndex.map { case (arg, position) =>
+      val param = params(position)
+      param.name -> arg
+    }.toMap
+  }
+
+  private def toScalaValue(expr: Expression): Any = expr match {
+    case literal: Literal => CatalystTypeConverters.convertToScala(literal.value, literal.dataType)
+    case _ => throw new AnalysisException(s"Cannot convert '$expr' to a Scala literal value")
+  }
+
+  private def resolveCatalog(nameParts: Seq[String]): (CatalogPlugin, Identifier) = {
+    val catalogManager = SparkSession.active.sessionState.catalogManager
+
+    if (nameParts.length == 1) {
+      return (catalogManager.currentCatalog, Identifier.of(catalogManager.currentNamespace, nameParts.head))
+    }
+
+    try {
+      val catalogName = nameParts.head
+      val procedureNameParts = nameParts.tail
+      (catalogManager.catalog(catalogName), toIdentifier(procedureNameParts))
+    } catch {
+      case _: CatalogNotFoundException =>
+        (catalogManager.currentCatalog, toIdentifier(nameParts))
+    }
+  }
+
+  private def toIdentifier(nameParts: Seq[String]): Identifier = {
```

Comment:
I could use `CatalogV2Implicits` and `LookupCatalog` here but I am not sure how safe it will be.